### PR TITLE
Feature/cam grid2obs priority

### PR DESCRIPTION
**Pull request by MarcelCaron-NOAA**
## Pull Request Testing ##

- [x] Describe testing already performed for this Pull Request:</br>

The full set of jobs has been running in the cron successfully for a week.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Similar to feature/CAM_precip, I recommend setting a few commands in the cron in two steps:  

(1) Replace {HOMEevs} in the following with the EVS directory that you are testing:

```
0 0 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswarw_grid2obs_prep.out {HOMEevs}/ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswarw_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswarwmem2_grid2obs_prep.out {HOMEevs}/ecf/cam/prep/jevs_hireswarwmem2_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswarwmem2_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswfv3_grid2obs_prep.out {HOMEevs}/ecf/cam/prep/jevs_hireswfv3_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswfv3_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hrrr_grid2obs_prep.out {HOMEevs}/ecf/cam/prep/jevs_hrrr_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hrrr_grid2obs_prep.out 2>&1
0 0 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_namnest_grid2obs_prep.out {HOMEevs}/ecf/cam/prep/jevs_namnest_grid2obs_prep.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_namnest_grid2obs_prep.out 2>&1
0 2 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswarw_grid2obs_stats.out {HOMEevs}/ecf/cam/stats/jevs_hireswarw_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswarw_grid2obs_stats.out 2>&1
0 2 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswarwmem2_grid2obs_stats.out {HOMEevs}/ecf/cam/stats/jevs_hireswarwmem2_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswarwmem2_grid2obs_stats.out 2>&1
0 2 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hireswfv3_grid2obs_stats.out {HOMEevs}/ecf/cam/stats/jevs_hireswfv3_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hireswfv3_grid2obs_stats.out 2>&1
0 2 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_hrrr_grid2obs_stats.out {HOMEevs}/ecf/cam/stats/jevs_hrrr_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_hrrr_grid2obs_stats.out 2>&1
0 2 * * * qsub -o /lfs/h2/emc/stmp/perry.shafran/jevs_namnest_grid2obs_stats.out {HOMEevs}/ecf/cam/stats/jevs_namnest_grid2obs_stats.ecf >> /lfs/h2/emc/stmp/perry.shafran/cron.out/test_jevs_namnest_grid2obs_stats.out 2>&1
```
(2) Then copy and paste the same code segment into your crontab.  It consists of five prep jobs run at 00Z every day, and five stats jobs run at 02Z each day.  The prep jobs should take about 1-2 minutes to complete.  The stats jobs should take around 2 hours to complete.

- [Yes] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

- [Yes] Do these updates/additions include sufficient testing updates? **[Yes or No]**

- [12/15/2022] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

**Comment by ShelleyMelchior-NOAA** [comment]
I cloned Marcel's fork and checked out his branch `feature/CAM_grid2obs_priority`.
I spot checked the 5 grid2obs prep ecf scripts, as in I ran them from command line to be sure they can run to completion before setting them up in cron. The following 5 ecf scripts ran to completion: 
`jevs_hireswarw_grid2obs_prep.ecf`
`jevs_hireswarwmem2_grid2obs_prep.ecf`
`jevs_hireswfv3_grid2obs_prep.ecf`
`jevs_hrrr_grid2obs_prep.ecf`
`jevs_namnest_grid2obs_prep.ecf`
All processes wrote to the same location, `COMOUT=/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/com/evs/v1.0/prep/cam/atmos/spc_otlk.20221219`

@MarcelCaron-NOAA , is that expected?

All 5 ecf scripts are now set up in cron to be initiated at 00:00Z.

**Comment by ShelleyMelchior-NOAA** [comment]
@MarcelCaron-NOAA , I tried to run `jevs_hireswarw_grid2obs_stats.ecf` but it fails because ${cyc} is not defined upstream of `JEVS_CAM_STATS`. Should I be exporting cyc from the ecf script? And if so, what should it be set to?

See line 2497 of the output log file:
`/lfs/h2/emc/stmp/shelley.melchior/jevs_hireswarw_grid2obs_stats.out`

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA the error is the result of the most recent merge of develop into this feature branch.  I just pushed a fix that should avoid the issue.  I was able to run the stats step successfully after this fix.

[Specifically, I'd kept calls to "setpdy.sh" to set the VDATE, but my jobs don't define VDATE this way, and so setpdy.sh would fail.  The fix allows both my method (without calling setpdy.sh) and the other developers' method of defining VDATE, and the method depends on the model being verified.]

Thanks.

**Comment by ShelleyMelchior-NOAA** [comment]
Excellent! I updated my clone of your fork and am now running `jevs_hireswarw_grid2obs_stats.ecf`. The job is still running, as expected, and output is being written to:
`COMOUT=/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/com/evs/v1.0/stats/cam/atmos`

**Comment by ShelleyMelchior-NOAA** [comment]
It finished. The output log fie: `/lfs/h2/emc/stmp/shelley.melchior/jevs_hireswarw_grid2obs_stats.out` for your review b/c I am not convinced this worked correctly.

**Comment by LoganDawson-NOAA** [comment]
@MarcelCaron-NOAA just want to check in on @ShelleyMelchior-NOAA's earlier comment about the prep jobs. Do those jobs do some preprocessing of each model's forecast files? Or is the main purpose to generate the SPC outlook and FireWx Nest mask files?

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA Yes it looks like the hireswarw job has run as expected.  Here is the output file for the hireswarw verification:
/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/com/evs/v1.0/stats/cam/atmos/hireswarw.20221220/evs.stats.hireswarw.atmos.grid2obs.v20221220.stat

It is about the same size as the ones I have been producing.  I skimmed a bit of the data in that file and they seem normal to me.  

The output log file you mention includes all previous runs, but output from the most recent run are located toward the bottom.  I'm hoping that the WCOSS2 scrubbers are able to clear them regularly, otherwise the cron job may need to be modified.

@LoganDawson-NOAA The prep step only gathers SPC outlook shapefiles, which are then processed in the stats step.  I was mistaken when I said they also gathered FireWx data files for masking.  There is no other processing done during the prep step (no preprocessing is done on the models' forecast files).

**Comment by ShelleyMelchior-NOAA** [comment]
I have tested all 10 of the prep and stats routines both as one-off command line runs and from cron (but not at the prescribed times in the PR instructions). 2022122500 and 2022122502 will be the first cron runs that will kick off as prescribed by the PR instructions. I will let them run for a few days over the weekend/holiday and Marcel can look at the output to confirm everything looks as expected.
`/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/com/evs/v1.0/`

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA I noticed some jobs worked and outputted normally (`ls /lfs/h2/emc/ptmp/shelley.melchior/EVS_out/com/evs/v1.0/stats/cam/atmos/*.20221224/*.stat`), and other more recent jobs have failed.  I found the reason for the failure, corrected the code, and pushed the change to Github.  I will monitor the output again once the PR test code has been updated, but in the meantime, since I have been running the same code concurrently on my cron, I will check the output from my cron jobs to see if the recent fix works.

**Comment by PerryShafran-NOAA** [comment]
@MarcelCaron-NOAA - if there are items missing from the authoritative run.ver, I should know so I can post them there.  Currently, the envir variable is there and so is the ccpa_ver.  We need to make sure that everything we need is in a single authoritative run.ver file.  

Recommend bringing the authoritative run.ver and testing would be good to be able to identify what else is missing so I can put it in there.  Feel free to check the authoritative run.ver vs yours, and if the authoritative has all the libraries needed, you should use it for your tests.  

NCO also won't allow any version numbers hardcoded anywhere other than the run.ver file, so we'll need to make sure we have all what we need in the single run.ver file.

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks Marcel! I will update my clone w/ your code changes and then update my crons. I will do this Thursday night or Friday when I can get back on VPN.

**Comment by MarcelCaron-NOAA** [comment]
@PerryShafran-NOAA Thanks, will do.  I just merged the new run.dev changes, and it looks like both ccpa_ver and envir are defined in the updated file, as you said.  With ccpa_ver and envir defined, the CAM verification jobs run correctly.

@ShelleyMelchior-NOAA sounds good!

File: ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf [hunk 1]
```diff
@@ -78,7 +78,6 @@ export MET_CONFIG="${MET_PLUS_PATH}/parm/met_config"
 export PYTHONPATH=$HOMEevs/ush/$COMPONENT:$PYTHONPATH
 
 # Developer Settings
-export envir=prod
 export DATA=/lfs/h2/emc/stmp/$USER/evs_test/$envir/tmp
```
Comment (ShelleyMelchior-NOAA) [inline ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf:81]: @MarcelCaron-NOAA  removing the definition of envir will break the definition of DATA in the next line. Should DATA be corrected?

Comment (ShelleyMelchior-NOAA) [reply ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf:81]: Oh! Nevermind ... I see that it is now exported from `run.ver`.

**Comment by ShelleyMelchior-NOAA** [comment]
I pulled the changes into my clone of Marcel's fork. I forced the crons to run at 16:40 for prep and 17:00 for stats on 20221230. The crons will be back on schedule for 00:00 and 02:00 for 20221231.

I also wrote that message on NYE and never hit the green "Comment" button for it to be published. My bad.

Anyway, the crons are running but I think two of the stats jobs got bit by the cactus nodes problem and were left in a hung state. I ran qdel to terminate those processes. @MarcelCaron-NOAA  please review my output.

**Comment by MarcelCaron-NOAA** [comment]
@ShelleyMelchior-NOAA Thanks for your comments over the holidays.  Looking at the output directory of the PR test jobs, it seems to me that the runs since 01/01/2023 have been completing successfully;  the output files match exactly what I've been producing at the same time and file sizes for each model are as expected (~14M for hireswarw/hireswarwmem2, ~16M for hireswfv3, ~65M for hrrr, ~35M for namnest).

I'm not sure why the Cactus issues haven't caused my cron jobs to hang too, but apart from the hung PR test jobs and missing output, the PR code seems to me to be working properly given the most recent output.

File: ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf (hunk 1 above)
Comment (MarcelCaron-NOAA) [reply ecf/cam/prep/jevs_hireswarw_grid2obs_prep.ecf:81]: Yes!  "envir" was added and then later removed from the ecf script in response to upstream changes in run.ver

**Comment by ShelleyMelchior-NOAA** [comment]
OK, sounds good, Marcel! I will proceed w/ merging this PR.

**Review by ShelleyMelchior-NOAA — APPROVED** [review]
This PR has been reviewed, tested, and is ready for merge into develop of the auth EVS repo.

**Pull request merged**